### PR TITLE
kubelet: fix mixing up runtime classes with runtime handlers

### LINKED ISSUE #123906: Issues with runtime handler supported feature lookup
There were a couple changes to merged in v1.30 to guard user namespaces & recursive read only mounts based on runtime self-reported support. When looking through the code, I noticed the following issues:

1. Lookup is done by `runtimeClassName`, rather than `runtimeHandler`:
    - https://github.com/kubernetes/kubernetes/blob/634fc1b4836b3a500e0d715d71633ff67690526a/pkg/kubelet/kubelet_pods.go#L2453
    - https://github.com/kubernetes/kubernetes/blob/634fc1b4836b3a500e0d715d71633ff67690526a/pkg/kubelet/kuberuntime/util/util.go#L104-L107
    - Should be using `LookupRuntimeHandler` https://github.com/kubernetes/kubernetes/blob/634fc1b4836b3a500e0d715d71633ff67690526a/pkg/kubelet/runtimeclass/runtimeclass_manager.go#L61
2. Version skew is not handled in the case where the container runtime doesn't provide any RuntimeHandlers, which results in a lot of log spam:
    ```
    Mar 12 19:17:29 kind-control-plane kubelet[719]: E0312 19:17:29.342999     719 kubelet_pods.go:2464] "unknown runtime class" runtimeClassName=""
    Mar 12 19:17:29 kind-control-plane kubelet[719]: E0312 19:17:29.347115     719 kubelet_pods.go:2464] "unknown runtime class" runtimeClassName=""
    Mar 12 19:17:30 kind-control-plane kubelet[719]: E0312 19:17:30.286822     719 kubelet_pods.go:2464] "unknown runtime class" runtimeClassName=""
    Mar 12 19:17:30 kind-control-plane kubelet[719]: E0312 19:17:30.286858     719 kubelet_pods.go:2464] "unknown runtime class" runtimeClassName=""
    Mar 12 19:17:30 kind-control-plane kubelet[719]: E0312 19:17:30.286864     719 kubelet_pods.go:2464] "unknown runtime class" runtimeClassName=""
    ```

/assign @giuseppe @AkihiroSuda
/cc @haircommander 

/kind bug regression
/milestone v1.30
/sig node

### PR DESCRIPTION
**Pull request by AkihiroSuda**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:
-  kubelet: fix mixing up runtime classes with runtime handlers
- kubelet: silence "unknown runtime class" errors when unsupported

#### Which issue(s) this PR fixes:

Fixes #123906
#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: pkg/kubelet/nodestatus/setters.go [hunk 1]
```diff
@@ -481,18 +483,41 @@ func GoRuntime() Setter {
...
+				klog.ErrorS(err, "Failed to look up the runtime handler", "runtimeClassName", class.Name)
+				continue
+			}
+			handler, ok := handlersMap[handlerName]
+			if !ok {
+				klog.ErrorS(nil, "Failed to look up the runtime handler", "runtimeClassName", class.Name,
+					"runtimeHandlerName", handlerName)
+				continue
+			}
 			node.Status.RuntimeClasses[i] = v1.NodeRuntimeClass{
```
Comment (AkihiroSuda) [inline pkg/kubelet/nodestatus/setters.go:517]: Wondering if we should have rather returned the list of the RuntimeHandlers, if we are still allowed to change the API

@thockin @haircommander @giuseppe

**Review by giuseppe — APPROVED** [review]
LGTM

File: pkg/kubelet/nodestatus/setters.go [hunk 2]
```diff
@@ -481,20 +483,49 @@ func GoRuntime() Setter {
...
+		}
+
+		classes, err := rcManager.Lister().List(labels.Everything())
+		if err != nil {
+			return fmt.Errorf("failed to get the list of the runtime classes: %w", err)
+		}
+		node.Status.RuntimeClasses = []v1.NodeRuntimeClass{}
+		handler, ok := handlersMap[""]
+		if !ok {
+			return fmt.Errorf("failed to look up the default runtime handler")
```
Comment (tallclair) [inline pkg/kubelet/nodestatus/setters.go:507]: How is this error handled? This case will be hit if it's using a CRI runtime that hasn't implemented the handlers yet.

File: pkg/kubelet/nodestatus/setters.go [hunk 3]
```diff
@@ -481,20 +483,49 @@ func GoRuntime() Setter {
...
-		for i, h := range handlers {
-			node.Status.RuntimeClasses[i] = v1.NodeRuntimeClass{
-				Name: h.Name,
+		handlers := handlersFunc()
+		handlersMap := make(map[string]kubecontainer.RuntimeHandler, len(handlers))
+		for _, h := range handlers {
+			handlersMap[h.Name] = h
+		}
+
+		classes, err := rcManager.Lister().List(labels.Everything())
```
Comment (tallclair) [inline pkg/kubelet/nodestatus/setters.go:500]: Does list return a stable order? If not, this list should probably be sorted to ensure that the list in the status is stable.

File: pkg/kubelet/nodestatus/setters.go [hunk 4]
```diff
@@ -481,20 +483,49 @@ func GoRuntime() Setter {
...
+			Name: "", // for the default runtime class
+			Features: &v1.NodeRuntimeClassFeatures{
+				RecursiveReadOnlyMounts: &handler.SupportsRecursiveReadOnlyMounts,
+			},
+		})
+
+		for _, class := range classes {
+			handler, ok = handlersMap[class.Handler]
+			if !ok {
+				klog.ErrorS(nil, "Failed to look up the runtime handler", "runtimeClassName", class.Name,
```
Comment (tallclair) [inline pkg/kubelet/nodestatus/setters.go:519]: This is not necessarily an error. A use case is that there is a dedicated node pool for running sandboxed containers, and a certain RuntimeClass/handler is only available in that node pool. In fact, this is a supported use case through the Scheduling mix-in on the RuntimeClass: https://github.com/kubernetes/api/blob/96558b97565e2a9bed0b5cb84db3b09b71d20172/node/v1/types.go#L65

File: pkg/kubelet/nodestatus/setters.go [hunk 5]
```diff
@@ -481,20 +483,49 @@ func GoRuntime() Setter {
...
+		handlersMap := make(map[string]kubecontainer.RuntimeHandler, len(handlers))
+		for _, h := range handlers {
+			handlersMap[h.Name] = h
+		}
+
+		classes, err := rcManager.Lister().List(labels.Everything())
+		if err != nil {
+			return fmt.Errorf("failed to get the list of the runtime classes: %w", err)
+		}
+		node.Status.RuntimeClasses = []v1.NodeRuntimeClass{}
```
Comment (tallclair) [inline pkg/kubelet/nodestatus/setters.go:504]: How is this API consumed? Or is it informational only? I wonder if the API should just be a list of RuntimeHandlers instead?

File: pkg/kubelet/kubelet_pods.go [hunk 6]
```diff
@@ -2445,23 +2445,32 @@ func (kl *Kubelet) cleanupOrphanedPodCgroups(pcm cm.PodContainerManager, cgroupP
...
+		return false
+	}
 	for _, h := range runtimeHandlers {
-		if h.Name == runtimeClassName {
+		if h.Name == runtimeHandlerName {
 			return h.SupportsRecursiveReadOnlyMounts
 		}
 	}
-	klog.ErrorS(nil, "unknown runtime class", "runtimeClassName", runtimeClassName)
+	klog.ErrorS(nil, "unknown runtime handler", "runtimeHandlerName", runtimeHandlerName)
```
Comment (tallclair) [inline pkg/kubelet/kubelet_pods.go:2473]: nit
```suggestion
	klog.ErrorS(nil, "Unknown runtime handler", "runtimeHandlerName", runtimeHandlerName)
```

**Review by ffromani — COMMENTED** [review]
LGTM

File: pkg/kubelet/runtimeclass/runtimeclass_manager.go [hunk 7]
```diff
@@ -76,3 +76,8 @@ func (m *Manager) LookupRuntimeHandler(runtimeClassName *string) (string, error)
 
 	return rc.Handler, nil
 }
+
+// Lister returns the runtime class lister.
+func (m *Manager) Lister() nodev1.RuntimeClassLister {
```
Comment (tallclair) [inline pkg/kubelet/runtimeclass/runtimeclass_manager.go:81]: Let's just expose a `List()` function that lists everything instead.

File: pkg/kubelet/kuberuntime/util/util_test.go [hunk 8]
```diff
@@ -223,7 +223,7 @@ func TestNamespacesForPod(t *testing.T) {
 		},
 	} {
 		t.Run(desc, func(t *testing.T) {
-			actual, err := NamespacesForPod(test.input, &kubecontainertest.FakeRuntimeHelper{})
+			actual, err := NamespacesForPod(test.input, &kubecontainertest.FakeRuntimeHelper{}, nil)
```
Comment (tallclair) [inline pkg/kubelet/kuberuntime/util/util_test.go:226]: This should have user namespace coverage.

**Review by tallclair — COMMENTED** [review]
Thanks for addressing this.

**Comment by tallclair** [comment]
/cc @thockin 

For the API discussion.

File: pkg/kubelet/nodestatus/setters.go (hunk 2 above)
Comment (AkihiroSuda) [reply pkg/kubelet/nodestatus/setters.go:507]: The code is removed

File: pkg/kubelet/nodestatus/setters.go (hunk 3 above)
Comment (AkihiroSuda) [reply pkg/kubelet/nodestatus/setters.go:500]: The code is removed

File: pkg/kubelet/nodestatus/setters.go (hunk 4 above)
Comment (AkihiroSuda) [reply pkg/kubelet/nodestatus/setters.go:519]: The code is removed

File: pkg/kubelet/kubelet_pods.go (hunk 6 above)
Comment (AkihiroSuda) [reply pkg/kubelet/kubelet_pods.go:2473]: done

File: pkg/kubelet/runtimeclass/runtimeclass_manager.go (hunk 7 above)
Comment (AkihiroSuda) [reply pkg/kubelet/runtimeclass/runtimeclass_manager.go:81]: The code is removed

File: pkg/kubelet/nodestatus/setters.go (hunk 1 above)
Comment (AkihiroSuda) [reply pkg/kubelet/nodestatus/setters.go]: Renamed to RuntimeHandlers

File: pkg/kubelet/kuberuntime/util/util_test.go (hunk 8 above)
Comment (AkihiroSuda) [reply pkg/kubelet/kuberuntime/util/util_test.go:226]: Yes, but seems beyond the scope of the PR.

cc @giuseppe @rata

Comment (rata) [reply pkg/kubelet/kuberuntime/util/util_test.go:226]: I'll open a follow-up PR for that, thanks!

File: staging/src/k8s.io/api/core/v1/types.go [hunk 9]
```diff
@@ -5925,11 +5925,11 @@ type NodeStatus struct {
 	// Status of the config assigned to the node via the dynamic Kubelet config feature.
 	// +optional
 	Config *NodeConfigStatus `json:"config,omitempty" protobuf:"bytes,11,opt,name=config"`
-	// The available runtime classes.
+	// The available runtime handlers.
 	// +featureGate=RecursiveReadOnlyMounts
 	// +optional
 	// +listType=atomic
-	RuntimeClasses []NodeRuntimeClass `json:"runtimeClasses,omitempty" protobuf:"bytes,12,rep,name=runtimeClasses"`
+	RuntimeHandlers []NodeRuntimeHandler `json:"runtimeHandlers,omitempty" protobuf:"bytes,12,rep,name=runtimeHandlers"`
```
Comment (liggitt) [inline staging/src/k8s.io/api/core/v1/types.go:5932]: this is brand new in 1.30? if so, it's important this field name change land before release

cc @thockin

**Comment by thockin** [comment]
API name change LGTM

/approve
/unhold

**Comment by thockin** [comment]
Who has the LGTM ?

**Comment by dims** [comment]
let's tag @mrunalp !

**Review by mrunalp — APPROVED** [review]
/lgtm

File: pkg/registry/core/node/strategy.go [hunk 10]
```diff
@@ -104,7 +104,7 @@ func dropDisabledFields(node *api.Node, oldNode *api.Node) {
 	}
 
 	if !utilfeature.DefaultFeatureGate.Enabled(features.RecursiveReadOnlyMounts) {
```
Comment (tallclair) [inline pkg/registry/core/node/strategy.go:106]: It seems weird that this is tied to the RecursiveReadOnlyMounts since it's also used by UserNamespaces.

Comment (AkihiroSuda) [reply pkg/registry/core/node/strategy.go:106]: The flag for UserNamespaces is still not included in the Kubernetes core API struct

**Pull request merged**